### PR TITLE
Add STABLE_PACKAGE_NO_IMPORT_UNSTABLE lint rule

### PR DESCRIPTION
**Pull request by bufdev**
I'm not yet sure if we want to add this into mainline. Opening this for discussion.

This is re: https://github.com/bufbuild/buf/issues/2986.

This adds a `STABLE_PACKAGE_NO_IMPORT_UNSTABLE` lint rule, with no category, that will produce a lint failure if an stable package imports a file with an unstable package. For example:

1. If `a.proto` imports `b.proto`, and `a.proto` has package `a.v1`:

- If `b.proto` has package `foo`, no failure - cannot determine package stability.
- If `b.proto` has package `foo.v1`, no failure - stable imports stable.
- If `b.proto` has package `foo.v1beta1`, failure - stable imports unstable.

2. If `a.proto` has package `a`:

- No failure ever - cannot determine package stability.

3. If `a.proto` has package `a.v1beta1`:

- No failure ever - package is unstable, does not matter what it imports.

If we do decide to go ahead with this, this PR:

- Needs some discussion as to the rule name.
- Needs testing.

### CODE REVIEW

File: private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go [hunk 1]
```diff
@@ -1029,6 +1029,32 @@ func checkServiceSuffix(add addFunc, service bufprotosource.Service, suffix stri
...
+		return nil
+	}
+	// Package is stable. Check imports.
+	for _, fileImport := range file.FileImports() {
+		packageVersion, ok := protoversion.NewPackageVersionForPackage(fileImport.File().Package())
+		if !ok {
+			continue
+		}
+		if packageVersion.StabilityLevel() != protoversion.StabilityLevelStable {
+			add(fileImport, fileImport.Location(), nil, `This file has package %q that represents a stable version, but it imports %q, which has package %q that represents an unstable version. Stable packages should not depend on unstable packages.`, file.Package(), fileImport.Import(), fileImport.File().Package())
```
Comment (akshayjshah) [inline private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go:1052]: Can we shorten the error text?

`This file is in stable package %q, so it should not depend on %q from unstable package %q.`

**Review by akshayjshah — COMMENTED** [review]
This feels like a valuable check to me!

Organizations with many existing violations of this rule may have a hard time adopting it, but that shouldn't stop us from offering it.

File: private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go (hunk 1 above)
Comment (noahseger) [reply private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go:1052]: 👍 @akshayjshah @bufdev 

I'm just doc golfing but less text is always nice`Stable package %q cannot import unstable file %q.`

File: private/bufpkg/bufcheck/buflint/testdata/lint_stable_package_import/api/v1/foo.proto [hunk 2]
```diff
@@ -0,0 +1,9 @@
+syntax = "proto3";
+
+package api.v1;
+
+import "api/v1alpha/bar.proto";
```
Comment (bufdev) [inline private/bufpkg/bufcheck/buflint/testdata/lint_stable_package_import/api/v1/foo.proto:5]: An import on a stable package, and a package with no stability level, should also be added (both should not result in failures)

**Review by bufdev — COMMENTED** [review]
Changelog needs to be updated

**Review by bufdev — COMMENTED** [review]
LGTM

**Pull request merged**